Add Helm chart for installing Tekton Operator

jacksgt opened the pull request:
Hello Tekton operators!

# Changes
As discussed in https://github.com/tektoncd/operator/issues/543  , the tekton-operator could really use a Helm chart for installation.
Helm is widely used in the Kubernetes ecosystem and allows for easy configuration of the installed components.
At the same time it gives much more flexibility when compared to installing from OperatorHub.

We are using an internal Helm chart for deploying the Operator, so I thought I'd share the chart here.

I have not added extensive documentation for it yet, but will do before merging this PR.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->

```release-note
Add Helm chart for installing Tekton Operator
```

dirien commented:
Cool finally we have a helm chart! Thanks @jacksgt!

sm43 commented:
Hi @jacksgt 
welcome to the tekton community 👋🏽 🙂
Could you please sign the Easy CLA? that would be required for getting this pr in😄

sm43 commented on chart/crds/openshift.yaml:
```diff
@@ -0,0 +1,186 @@
+{{- if .Values.openshift.enabled -}}
+---
+apiVersion: apiextensions.k8s.io/v1
```
don't have any idea about how helm works
but is there any way we can refer this form the CRD yaml?

instead of adding CRD definition here, do we have a way to import the yaml?
so in future any changes there will automatically be added in charts 😃

jacksgt replied:
We should definitely have a way to keep the CRDs in-sync!
We cannot reference anything "outside" the Helm chart, so we'll need to copy the CRDs inside this directory - this should not be difficult to do as part of the release process, though I think it makes sense to handle this in a separate PR :-)

vdemeester replied:
Yes, we need to automate that, but it could be done in a follow-up 😉

sm43 commented:
>I have not added extensive documentation for it yet, but will do before merging this PR

yep documentation would be useful to understand and test it out 👍🏽 🙂

jacksgt commented:
> welcome to the tekton community 👋🏽 slightly_smiling_face

Hi, thanks!

> Could you please sign the Easy CLA? that would be required for getting this pr in

Sure, done!

> yep documentation would be useful to understand and test it out 👍🏽 slightly_smiling_face

I have added a README. Let me know if the instructions are clear for you.

sm43 commented on chart/templates/service.yaml:
```diff
@@ -0,0 +1,57 @@
+  annotations:
+    networkoperator.openshift.io/ignore-errors: ""
+  {{- end }}
+  labels:
```
```suggestion
```

jacksgt replied:
Whoops, not sure how that slipped through, thanks!

sm43 commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,439 @@
+  name: {{ include "tekton-operator.fullname" . }}
+  labels:
+    {{- include "tekton-operator.labels" . | nindent 4 }}
+rules:
```
this rules are for service account for kubernetes platform

for openshift the rules are bit different
https://github.com/tektoncd/operator/blob/main/config/openshift/base/role.yaml

so currently on openshift helm chart is failing

sm43 commented:
another reason it is failing on openshift is https://github.com/tektoncd/operator/pull/643

sm43 commented on chart/README.md:
````diff
@@ -0,0 +1,62 @@
+
+```sh
+git clone https://github.com/tektoncd/operator.git
+helm install tekton-operator \
````
on kubernetes CRDs are not getting installed 🤔

sm43 commented:
@jacksgt https://github.com/tektoncd/operator/blob/b5ce705a074e79beceaf35789955db243a4b965b/config/base/tekton-config-defaults.yaml#L18 configmap is not getting created on openshift?

jacksgt commented:
> @jacksgt
> 
> https://github.com/tektoncd/operator/blob/b5ce705a074e79beceaf35789955db243a4b965b/config/base/tekton-config-defaults.yaml#L18
> configmap is not getting created ? on both k8s and openshift

Those settings are directly injected as environment variables into the Tekton operator: https://github.com/tektoncd/operator/pull/627/files#diff-2c6c478bf17d4896de57d0def131d88f926a02e3a0159954b1e1457a19b78bebR54

nikhil-thomas commented:
@jacksgt 💯 👍  for this pr. 
I'm testing this right now. I shall come back with my questions and comments in a couple of days. 🧑‍💻

please excuse the delay. we got a bit busy trying to handle the new additions like TektonHub and TektonChains in operator. 🕺

sm43 commented:
@jacksgt sorry for the delay in reviews.
I tried but its failed on k8s and openshift.
can you update it to latest operator version and update the chart
I will try it and then we can merge.
We will also need to think how to manage the chart with new release?
some kind of automation ..?

jacksgt commented:
Hi @sm43 , sorry for the late reply, I finally got around to addressing your feedback.

> I tried but its failed on k8s and openshift.

There was a minor issue in the YAML names. I fixed it and verified that it deploys on k8s and openshift.

> can you update it to latest operator version and update the chart

Yes, done.

> I will try it and then we can merge.

Please give it another go :-)

> We will also need to think how to manage the chart with new release?
> some kind of automation ..?

Yes, I can address this in a follow-up MR. I guess you have some kind of script for creating a release, we just need to add a few commands there (for updating the appVersion and chart version).

vdemeester reviewed (approved):
/ok-to-test

nikhil-thomas commented:
/retest

as https://github.com/tektoncd/operator/pull/769 has merged, ci should pass

sm43 commented:
/retest

jacksgt commented:
Hello folks, the tests passed, is there anything else left to do for this PR? :-)

vdemeester commented:
@jacksgt can you squash your commits ?
@sm43 can you take a last look ?

Once it is squash, I am more than ok to get this merge, and iterate over it if there is some bugs, etc.

jacksgt commented:
> Thanks! I just tried, there are couple of things to fix

Thanks for taking the time to check again!

> 
>     * in 0.56.0 there are 2 new CRDs added, [hub](https://github.com/tektoncd/operator/blob/main/config/base/300-operator_v1alpha1_hub_crd.yaml) and [chains](https://github.com/tektoncd/operator/blob/main/config/base/300-operator_v1alpha1_chain_crd.yaml)

Thanks for the hint! I synced the CRDs with the YAMLs from release 0.57
> 
>     * the operator deployment are created with the tag `v0.56.0` but it should be `0.56.0`

This seems to be an inconsistency with the 0.56 release.
I bumped the version to `v0.57.0` (and also rebased on main), so now it should be fine.

I'll make sure to submit a follow-up MR that will automatically sync the CRDs for each new release and bump the chart version.

sm43 commented:
for 0.57.0 you will need to include https://github.com/tektoncd/operator/commit/9819f1d55335b0d445c3a36d126ff1c1dd142a4e this change too ! then we are good to merge 🙃

jacksgt commented:
> for 0.57.0 you will need to include [9819f1d](https://github.com/tektoncd/operator/commit/9819f1d55335b0d445c3a36d126ff1c1dd142a4e) this change too ! then we are good to merge upside_down_face

Ohhh thanks for the hint, very kind of you!
I missed that change.

I updated the chart accordingly.

sm43 commented:
/test pull-tekton-operator-go-coverage

vdemeester commented:
/retest
/lgtm
